File: total-cost-to-hire-k-workers.py
'''
https://leetcode.com/problems/total-cost-to-hire-k-workers

1. min heap
time: nlogn
space: o(n)
'''
class Solution:
    def totalCost(self, costs: List[int], k: int, candidates: int) -> int:
        q1, q2 = [], []
        l, r = 0, len(costs) - 1
        res = 0
        while k > 0:
            while len(q1) < candidates and l <= r:
                heappush(q1, costs[l])                
                l += 1
            while len(q2) < candidates and l <= r:
                heappush(q2, costs[r])                
                r -= 1
            a = q1[0] if q1 else float(inf)
            b = q2[0] if q2 else float(inf)
            if a <= b:
                res += a
                heappop(q1)
            else:
                res += b
                heappop(q2)
            k -= 1
        return res
        

File: leetcode/leetcode_cpp/total-cost-to-hire-k-workers.cpp
/*
https://leetcode.com/problems/total-cost-to-hire-k-workers

1. min heap
time: nlogn
space: o(n)
*/
class Solution {
public:
    long long totalCost(vector<int>& costs, int k, int candidates) {
        priority_queue<int, vector<int>, greater<int>> q1, q2;
        int l = 0, r = costs.size()-1;
        long long res = 0;
        while (k > 0) {
            while (q1.size() < candidates && l<=r) q1.push(costs[l++]);
            while (q2.size() < candidates && l<=r) q2.push(costs[r--]);
            auto a = q1.size() ? q1.top() : INT_MAX;
            auto b = q2.size() ? q2.top() : INT_MAX;
            if (a <= b) {
                res += a;
                q1.pop();
            }
            else {
                res += b;
                q2.pop();
            }
            k -= 1;
        }
        return res;
    }
};